aten/src/ATen/core/thread_pool.cpp

```diff
@@ -71,12 +71,25 @@ void ThreadPool::workOnTasksUntilCompleted(
     return;
   }
   std::condition_variable finished;
-  future->addCallback([&] { finished.notify_all(); });
+  bool fired = false;
+  future->addCallback([&] {
+    std::unique_lock<std::mutex> future_lock(future->get_mutex());
+    finished.notify_all();
+    fired = true;
+  });
 
   std::unique_lock<std::mutex> future_lock(future->get_mutex());
-  while (!future->completed()) {
-    finished.wait(future_lock);
-  }
+  do {
+    // Two corner cases:
+    // (1) force to run the loop at least once in case the above callback
+    // is invoked after the current thread has completed,
+    // which will dispose "finished".
```

Comment:
I didn't understand this comment upon reading it.

What I think you are trying to say, is that there is a race between when the thread that calls `markCompleted` on the future releases the lock, and when the callback is run (because the callback is run without synchronization and has to reacquire the lock). So the order of events is something like:

1. We add callback to the future.
2. In another thread, it is marked completed. We mark the future as completed,
3. Main thread grabs the lock and does... something?

But what I still don't understand what the point of forcing the loop to run once is. For example, I think you could have fixed the bug simply by adding a lock acquire inside the callback. Could you tell me where I have reasoned incorrectly?